### PR TITLE
Update{Occurrence,Note} functionality with FieldMask support

### LINKED ISSUE #312: Implement Update{Note,Occurrence} operations using FieldMask
## Expected Behavior
Merge and update the fields in Note and Occurrence using the provided FieldMask.

## Actual Behavior
Currently, not implemented.

## Additional info
see `go/v1beta1/storage`

### PR DESCRIPTION
**Pull request by ItsMeWithTheFace**
Fixes #312.

I actually decided to use `github.com/mennanov/fieldmask-utils` for managing the field mask operations until golang has a native solution in their protobuf library. It's a well-tested package that [others have had success with](https://github.com/golang/protobuf/issues/398), but let me know if I should use something else.

I'm open to any comments or concerns!

### CODE REVIEW

**Comment by ItsMeWithTheFace** [comment]
I've been running into some weird build issues last night which started after I added new packages in (and had a cyclic dependency, which should be fixed now though). Is there a specific procedure you guys follow to vendor new packages? The `go.mod` isn't in the root of the project so I guessed things would be done differently

**Comment by clrprod** [comment]
re: adding new packages
The go.mod file in the tools directory is a bit deceptive. It's used to install some proto tools as binaries. If you need to add a new package, I believe we use dep to manage the vendor directory. There are two PRs to move over to go modules, but neither have been accepted yet. In the short term I would encourage you to use dep to add your new dependency to the vendor folder.

File: tools/go.mod [hunk 1]
```diff
@@ -5,3 +5,7 @@ go 1.12
 require github.com/golang/protobuf v1.3.1
 
 require github.com/grpc-ecosystem/grpc-gateway v1.9.0
+
+require github.com/imdario/mergo v0.3.7
```
Comment (clrprod) [inline tools/go.mod]: We don't use go modules, you shouldn't touch this file unless you're trying to pull down a new binary tool.

Comment (ItsMeWithTheFace) [reply tools/go.mod]: Sounds good!

**Comment by ItsMeWithTheFace** [comment]
> Also we should probably add some tests for this new update behaviour, a table with a few different starting occurrences, update masks, and output occurrences (and notes) would probably be sufficient

For this, I could add unit tests to the new functions I built and update the existing `Update{Occurrence, Notes}` tests as integration tests?

**Comment by aysylu** [comment]
> > Also we should probably add some tests for this new update behaviour, a table with a few different starting occurrences, update masks, and output occurrences (and notes) would probably be sufficient
> 
> For this, I could add unit tests to the new functions I built and update the existing `Update{Occurrence, Notes}` tests as integration tests?

That'd be great, thanks!

**Comment by aysylu** [comment]
Hey @ItsMeWithTheFace, the status of the PR is shown as `Draft`, which means that it's not ready for full review yet? Feel free to ping this thread when you'd like us to take another look. Thanks!

**Comment by ItsMeWithTheFace** [comment]
Yeah I'm currently working on a more robust implementation that I'll push out very soon. I'll let you guys know once that's done :)

**Comment by ItsMeWithTheFace** [comment]
@aysylu, @vtsao, @clrprod, I'm very close to a full review (just need to sort out this locking issue in `memstore.go`) but feel free to leave any feedback on this PR!

**Comment by ItsMeWithTheFace** [comment]
Everything should be done now! Ready for review

**Comment by aysylu** [comment]
Hi @ItsMeWithTheFace, thanks for your contribution! I apologize for the delay in the review, I'll be able to get to it next week. Thanks for your patience!

File: go/v1beta1/storage/pgsqlstore.go [hunk 2]
```diff
@@ -241,10 +241,19 @@ func (pg *PgSQLStore) DeleteOccurrence(ctx context.Context, pID, oID string) err
 // UpdateOccurrence updates the existing occurrence with the given projectID and occurrenceID
 func (pg *PgSQLStore) UpdateOccurrence(ctx context.Context, pID, oID string, o *pb.Occurrence, mask *fieldmaskpb.FieldMask) (*pb.Occurrence, error) {
 	o = proto.Clone(o).(*pb.Occurrence)
-	// TODO(#312): implement the update operation
-	o.UpdateTime = ptypes.TimestampNow()
 
-	result, err := pg.DB.Exec(updateOccurrence, pID, oID, proto.MarshalTextString(o))
+	oldOcc, err := pg.GetOccurrence(ctx, pID, oID)
+	if err != nil {
+		return nil, status.Error(codes.Internal, "Failed to retrieve Occurrence to update")
```
Comment (vtsao) [inline go/v1beta1/storage/pgsqlstore.go]: Just return the error directly.

**Comment by mickep76** [comment]
I wrote this for a CRUD back-end it's quite small and specific to get either map[string]interface{} or JSON.

https://github.com/mickep76/update_mask

**Comment by aysylu** [comment]
Closing this due to inactivity since September 2019. Please feel free to open a new PR with the comments addressed from this PR.

**Pull request closed without merging**